### PR TITLE
Add support for DAC.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

**Comment by floitsch** [comment]
There are a few TODOs where I would appreciate feedback.

File: lib/gpio/dac.toit [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  /** Phase shift constant of the cosine wave generator: 0° */
+  static COSINE_WAVE_PHASE_0/int ::= 2
+  /** Phase shift constant of the cosine wave generator: 180° */
+  static COSINE_WAVE_PHASE_180/int ::= 3
+
+  pin/Pin
+  resource_ := ?
+
+  /**
+  Initializes a Dac channel.
```
Comment (kasperl) [inline lib/gpio/dac.toit:49]:
```suggestion
  Initializes a DAC channel.
```

File: src/resources/dac_esp32.cc [hunk 2]
```diff
@@ -0,0 +1,236 @@
+// Copyright (C) 2021 Toitware ApS.
```
Comment (kasperl) [inline src/resources/dac_esp32.cc:1]:
```suggestion
// Copyright (C) 2022 Toitware ApS.
```

File: src/resources/dac_esp32.cc [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+static constexpr int kDacMinFrequency = 130;
+static constexpr int kDacMaxFrequency = 5500;
+
+#ifdef CONFIG_IDF_TARGET_ESP32
+
+static dac_channel_t get_dac_channel(int pin) {
+  switch (pin) {
+    case 25: return DAC_CHANNEL_1;
+    case 26: return DAC_CHANNEL_2;
+    default: return dac_channel_t(-1);
```
Comment (kasperl) [inline src/resources/dac_esp32.cc:45]: Is this a cast? Here and in a number of other places. Use static_cast?

File: src/resources/dac_esp32.cc [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+PRIMITIVE(use) {
+  ARGS(DacResourceGroup, group, int, pin, uint8, initial_value);
+
+  dac_channel_t channel = get_dac_channel(pin);
+  // TODO(florian): The compiler complains that the result of the comparison
+  // is always false (`-Wtautological-constant-out-of-range-compare`). I believe we use
+  // this in many other places. Do we care?
+  if (channel == -1) INVALID_ARGUMENT;
```
Comment (kasperl) [inline src/resources/dac_esp32.cc:164]: Cast -1 to dac_channel_t first?

File: src/resources/dac_esp32.cc [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+
+  // TODO(florian): the AdcState is allocated with a `HeapTagScope`. Do we need
+  // to do the same here?
+  DacResource* resource = _new DacResource(group, channel);
+  if (resource == null) MALLOC_FAILED;
+  // TODO(florian): if we fail, do we need to release the resource that isn't
```
Comment (kasperl) [inline src/resources/dac_esp32.cc:173]: Yes, you need to delete the resource before returning prior to registering the resource.

File: lib/gpio/dac.toit (hunk 1 above)
Comment (floitsch) [reply lib/gpio/dac.toit]: done.

File: src/resources/dac_esp32.cc (hunk 2 above)
Comment (floitsch) [reply src/resources/dac_esp32.cc]: done.

File: src/resources/dac_esp32.cc (hunk 3 above)
Comment (floitsch) [reply src/resources/dac_esp32.cc]: Not a cast. A "constructor" call.
Replaced with constexpr that had a cast. Now there is a `kInvalidChannel`.
done.

File: src/resources/dac_esp32.cc (hunk 4 above)
Comment (floitsch) [reply src/resources/dac_esp32.cc]: Using `kInvalidChannel` now.

File: src/resources/dac_esp32.cc (hunk 5 above)
Comment (floitsch) [reply src/resources/dac_esp32.cc]: done.

File: src/resources/dac_esp32.cc [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    unuse_cosine();
+  }
+
+  dac_channel_t channel() const { return _channel; }
+
+  esp_err_t use_cosine() {
+    esp_err_t err = ESP_OK;
+    if (_uses_cosine) return err;
+    _uses_cosine = true;
+    // TODO(florian): this needs to be protected in a lock.
```
Comment (floitsch) [inline src/resources/dac_esp32.cc:100]: @kasperl : would appreciate help in solving this TODO.

**Comment by floitsch** [comment]
I'm using the global mutex now.

File: src/resources/dac_esp32.cc [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+static dac_channel_t get_dac_channel(int pin) {
+  return kInvalidChannel;
+}
+
+#endif
+
+class DacResourceGroup;
+
+static int _cosine_user_count = 0;
```
Comment (kasperl) [inline src/resources/dac_esp32.cc:82]: We typically don't make these private (staring with _).

File: src/resources/dac_esp32.cc [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+  TAG(DacResource);
+  DacResource(DacResourceGroup* group, dac_channel_t channel);
+
+  virtual ~DacResource() override {
+    unuse_cosine();
+  }
+
+  dac_channel_t channel() const { return _channel; }
+
+  esp_err_t use_cosine() {
```
Comment (kasperl) [inline src/resources/dac_esp32.cc:95]: I would probably not have these in the class definition (use_cosine / unuse_cosine). No need to have them inlined.

**Review by kasperl — APPROVED** [review]
Looks good to me. Also with the resource mutex instead of the global one.

**Pull request merged**